src/components/data-exploration/embedding/Embedding.jsx

```diff
@@ -84,6 +92,20 @@ const Embedding = (props) => {
     }
   }, []);
 
+  useEffect(() => {
+    if (!sampleKeys?.length) return;
+
+    const allCellIds = sampleKeys
+      .reduce((acc, sampleKey) => [...acc, ...cellSetProperties[sampleKey].cellIds], [])
+      .filter((cellId) => cellId !== undefined);
+
+    const { forwardMap, reverseMap } = generateCellIdMapping(allCellIds);
+
+    setGetCellSetCellId(() => (embeddingCellIdx) => reverseMap[embeddingCellIdx]);
+    setGetEmbeddingCellIdx(() => (cellSetCellId) => forwardMap[cellSetCellId]);
+    setCellIndexMapping(forwardMap);
+  }, [sampleKeys?.length]);
```

Comment:
why is this useEffect necessary?

Reply:
This `useEffect` block wraps the code which generates the mapping between cellIds that we use in the platform vs the cellIds that we is used for the embedding. The block depends on the number of samples - which shouldn't change. This ensures that the mapping is only calculated once per page render.